core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractFlattenLastNode.java

```diff
@@ -0,0 +1,166 @@
+package org.optaplanner.constraint.streams.bavet.common;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.function.Function;
+
+public abstract class AbstractFlattenLastNode<InTuple_ extends Tuple, OutTuple_ extends Tuple, EffectiveItem_, FlattenedItem_>
+        extends AbstractNode
+        implements TupleLifecycle<InTuple_> {
+
+    private final int flattenLastStoreIndex;
+    private final Function<EffectiveItem_, Iterable<FlattenedItem_>> mappingFunction;
+    /**
+     * Calls for example {@link AbstractScorer#insert(Tuple)}, and/or ...
+     */
+    private final TupleLifecycle<OutTuple_> nextNodesTupleLifecycle;
+    private final Queue<OutTuple_> dirtyTupleQueue = new ArrayDeque<>(1000);
+
+    protected AbstractFlattenLastNode(int flattenLastStoreIndex,
+            Function<EffectiveItem_, Iterable<FlattenedItem_>> mappingFunction,
+            TupleLifecycle<OutTuple_> nextNodesTupleLifecycle) {
+        this.flattenLastStoreIndex = flattenLastStoreIndex;
+        this.mappingFunction = Objects.requireNonNull(mappingFunction);
+        this.nextNodesTupleLifecycle = Objects.requireNonNull(nextNodesTupleLifecycle);
+    }
+
+    @Override
+    public void insert(InTuple_ tuple) {
+        Object[] tupleStore = tuple.getStore();
+        if (tupleStore[flattenLastStoreIndex] != null) {
+            throw new IllegalStateException("Impossible state: the input for the tuple (" + tuple
+                    + ") was already added in the tupleStore.");
+        }
+        List<OutTuple_> outTupleList = new ArrayList<>();
+        for (FlattenedItem_ item : mappingFunction.apply(getEffectiveFactIn(tuple))) {
+            addTuple(tuple, item, outTupleList);
+        }
+        if (!outTupleList.isEmpty()) {
+            tupleStore[flattenLastStoreIndex] = outTupleList;
+        }
+    }
+
+    private void addTuple(InTuple_ originalTuple, FlattenedItem_ item, List<OutTuple_> outTupleList) {
+        OutTuple_ tuple = createTuple(originalTuple, item);
+        outTupleList.add(tuple);
+        tuple.setState(BavetTupleState.CREATING);
+        dirtyTupleQueue.add(tuple);
+    }
+
+    protected abstract OutTuple_ createTuple(InTuple_ originalTuple, FlattenedItem_ item);
+
+    @Override
+    public void update(InTuple_ tuple) {
```

Comment:
It is questionable if all this work can possibly be faster than just remove+insert.